zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxn.java

```diff
@@ -1121,6 +1125,9 @@ private void startConnect(InetSocketAddress addr) throws IOException {
                     LOG.warn("Unexpected exception", e);
                 }
             }
+            if (!state.isAlive()) {
+                throw new RuntimeException("Already closed");
+            }
```

Comment:
Just as you said, this issue happens when network is broken, so `IOException` is better? e.g. 
`throw new IOException("Socket has already closed when reconnecting to "+ addr);`

Reply:
I prefer to RuntimeException. In my opinion, IOException indicates the actual physical network is broken. But I think it is not proper to throw IOException here. Here it is software issue that misses the state check. The direct reason is not a real physical network broken.  I think that throwing of IOException here would confuse the users. So I still prefer to use RuntimeException here.

Reply:
`RuntimeException` indicates unrecoverable programming error - such as illegal arguments or unrecoverable errors. Our case here is not that dramatic, it's recoverable by throwing an IOException if the state is `CLOSED` and it's being set to another value. `CLOSED` state will signal `SenderThread` to terminate and the benefit is we will be able to clean up resources - like the sockets sender thread held, and more importantly, the event thread will also be properly shutdown. 
And since we are explicitly programming in this issue to prevent this error, it does not sound like a logic error to me that worth to throw a RuntimeException.

Reply:
Both of you convince me. Now I throw IOException in SendThread.setZkState. Please check latest patch